js/plugins/drupalentity/plugin.js

```diff
@@ -8,27 +8,25 @@
   "use strict";
 
   CKEDITOR.plugins.add('drupalentity', {
-    // This plugin requires the Widgets System defined in the 'widget' plugin.
     requires: 'widget',
 
-    // The plugin initialization logic goes inside this method.
     beforeInit: function (editor) {
       // Configure CKEditor DTD for custom drupal-entity element.
       // @see https://www.drupal.org/node/2448449#comment-9717735
-      var dtd = CKEDITOR.dtd, tagName;
-      dtd['drupal-entity'] = {'#': 1};
-      // Register drupal-entity element as allowed child, in each tag that can
-      // contain a div element.
-      for (tagName in dtd) {
-        if (dtd[tagName].div) {
-          dtd[tagName]['drupal-entity'] = 1;
-        }
-      }
-
+      CKEDITOR.dtd['drupal-entity'] = {'#': 1};
+      CKEDITOR.dtd.$block['drupal-entity'] = 1;
+      CKEDITOR.dtd.$blockLimit['drupal-entity'] = 1;
+      // @todo It should be possible to display embedded entities inline...
+      //CKEDITOR.dtd.$inline['drupal-entity'] = 1;
+      CKEDITOR.dtd.$nonEditable['drupal-entity'] = 1;
+      CKEDITOR.dtd.$object['drupal-entity'] = 1;
```

Comment:
Removing the above code (that loop) isn't correct. In the DTD you need to define a couple of things:

1. What your tag can contain (you did this with `CKEDITOR.dtd['drupal-entity'] = {'#': 1};` – it means that your tag can only contain a text).
2. Then, you need to define what other elements can be parents of your element. That was what the loop did. It said that every element that can contain a `div` will be able to contain your element (so, in other words – treat your element like a `div`).
3. Finally, there are some special objects which define additional semantics. The `$block` contains elements that should be treated as blocks (like `<p>`), `$blockLimit` is a block which cannot be freely split (like `<td>`) – to these two you should add your `drupal-entity`.

And that should be all, if your tag is later replaced by some widget. If it's not, then... well – you shouldn't be using non-HTML tags at all :P.